native: provide socket-based ZEP device

miri64 opened the pull request:
This is aiming to replace [GNRC ZEP] and allow platform independent IEEE 802.15.4 traffic on native (as opposed to #5582).

I tested it already with [zep_mesh], a very simple network simulator I specifically wrote for ZEP (and RIOT, though I tried to keep it generalized):

![zep_showcase](https://cloud.githubusercontent.com/assets/675644/20311724/25e29708-ab51-11e6-9501-3b51ef795f05.png)

This is more or less done, but there are still some minor stuff I might add/change on a later point:

- allow for multiple ZEP interfaces (see #2195, I'm planning to adopt this PR in the near future)
- filter channels, wrong FCS, etc.

Depends on ~~#6112~~ (merged).

[GNRC ZEP]: https://github.com/RIOT-OS/RIOT/tree/master/sys/net/gnrc/application_layer/zep
[zep_mesh]: https://github.com/miri64/zep_mesh

miri64 commented:
(Just in case you are wondering why the EUI-64 is gradually growing with the nodes: it's generated from the sockets sockaddr. I took great care that the local port part is in the region were the last two bytes of the EUI-64 are and also used those two bytes for the short address).

miri64 commented:
I implemented this primarily to test my new NDP implementation I'm going to finally start in a few days.

LudwigKnuepfer commented on cpu/native/socket_zep/socket_zep.c:
```diff
@@ -0,0 +1,424 @@
+            return (memcmp(dst_addr, ieee802154_addr_bcast, dst_len) != 0) &&
+                   (memcmp(dst_addr, dev->netdev.short_addr, dst_len) != 0);
+        default:
+            return false;    /* better save than sorry ;-) */
```
`s/save/safe/` I guess

LudwigKnuepfer commented on cpu/native/socket_zep/socket_zep.c:
```diff
@@ -0,0 +1,424 @@
+            if (max_len != sizeof(uint16_t)) {
+                return -EOVERFLOW;
+            }
+            *v = SOCKET_ZEP_FRAME_PAYLOAD_LEN;
```
you might want to check for `v != NULL` before...

miri64 replied:
I think we don't do that anywhere else, but will do.

miri64 replied:
Done in https://github.com/RIOT-OS/RIOT/pull/6121/files#diff-01bcdc6a0bf2aa346059e749a7178abaR309 btw ;-)

miri64 commented:
Addressed comments

miri64 commented:
I found the following patch useful for testing BTW ;-)

```diff
diff --git a/boards/native/Makefile.dep b/boards/native/Makefile.dep
index 88151e9..0991015 100644
--- a/boards/native/Makefile.dep
+++ b/boards/native/Makefile.dep
@@ -1,5 +1,5 @@
 ifneq (,$(filter netdev_default gnrc_netdev_default,$(USEMODULE)))
-  USEMODULE += netdev2_tap
+  USEMODULE += socket_zep
 endif
 
 ifneq (,$(filter socket_zep,$(USEMODULE)))
diff --git a/boards/native/Makefile.include b/boards/native/Makefile.include
index 01fc119..232dc26 100644
--- a/boards/native/Makefile.include
+++ b/boards/native/Makefile.include
@@ -88,11 +88,11 @@ endif
 export LINKFLAGS += -ffunction-sections
 
 # set the tap interface for term/valgrind
-ifneq (,$(filter netdev_default gnrc_netdev_default,$(USEMODULE)))
-       export PORT ?= tap0
-else
-       export PORT =
-endif
+# ifneq (,$(filter netdev_default gnrc_netdev_default,$(USEMODULE)))
+#      export PORT ?= tap0
+# else
+#      export PORT =
+# endif
 
 export TERMFLAGS := $(PORT) $(TERMFLAGS)
 
```

(will provide a proper test application ASAP)

miri64 commented:
I found the following patch useful for testing BTW ;-)

```diff
diff --git a/boards/native/Makefile.dep b/boards/native/Makefile.dep
index 88151e9..0991015 100644
--- a/boards/native/Makefile.dep
+++ b/boards/native/Makefile.dep
@@ -1,5 +1,5 @@
 ifneq (,$(filter netdev_default gnrc_netdev_default,$(USEMODULE)))
-  USEMODULE += netdev2_tap
+  USEMODULE += socket_zep
 endif
 
 ifneq (,$(filter socket_zep,$(USEMODULE)))
diff --git a/boards/native/Makefile.include b/boards/native/Makefile.include
index 01fc119..232dc26 100644
--- a/boards/native/Makefile.include
+++ b/boards/native/Makefile.include
@@ -88,11 +88,11 @@ endif
 export LINKFLAGS += -ffunction-sections
 
 # set the tap interface for term/valgrind
-ifneq (,$(filter netdev_default gnrc_netdev_default,$(USEMODULE)))
-       export PORT ?= tap0
-else
-       export PORT =
-endif
+# ifneq (,$(filter netdev_default gnrc_netdev_default,$(USEMODULE)))
+#      export PORT ?= tap0
+# else
+#      export PORT =
+# endif
 
 export TERMFLAGS := $(PORT) $(TERMFLAGS)
 
```

(will provide a proper test application ASAP)

kaspar030 commented:
Very nice! I won't find time to test, though...

kaspar030 commented on cpu/native/socket_zep/socket_zep.c:
```diff
@@ -0,0 +1,424 @@
+
+    if (getpeername(dev->sock_fd, (struct sockaddr *)&ss_array, &ss_len) < 0) {
+        perror("ZEP: Unable to retrieve remote address");
+        return;
```
Here and below: the socket is connected at this point, should it be closed?

miri64 replied:
I think that's a remnant from when I played around with unbound/unconnected sockets. Will do.

miri64 commented:
Provided tests.

miri64 commented:
Rebased to current master, since #6112 got merged => Removed Waiting for other PR label. Also removed WIP label. Those missing features can well be done in a follow-up.

PeterKietzmann commented:
@miri64 can you give a hint about how to use this, e.g. with `gnrc_networking`? I applied your patch and indeed, a `gnrc_socket_zep` thread appeared but unfortunately the `gnrc_netdev2_tap` interface is still predominant. Or am I mistaken in using `gnrc_networking` for testing this PR?

```
> ps  
ps
	pid | name                 | state    Q | pri | stack ( used) | base       | current    
	  - | isr_stack            | -        - |   - |  8192 (   -1) |  0x808a480 |  0x808a480
	  1 | idle                 | pending  Q |  15 |  8192 (  420) |  0x8088200 |  0x808a074 
	  2 | main                 | running  Q |   7 | 12288 ( 3124) |  0x8085200 |  0x8088074 
	  3 | pktdump              | bl rx    _ |   6 | 12288 (  976) |  0x8092500 |  0x8095374 
	  4 | 6lo                  | bl rx    _ |   3 |  8192 (  976) |  0x80a0ac0 |  0x80a2934 
	  5 | ipv6                 | bl rx    _ |   4 |  8192 ( 1424) |  0x809be40 |  0x809dcb4 
	  6 | udp                  | bl rx    _ |   5 |  8192 (  976) |  0x8095540 |  0x80973b4 
	  7 | gnrc_netdev2_tap     | bl rx    _ |   2 |  8192 ( 2356) |  0x8097c40 |  0x8099ab4 
	  8 | gnrc_socket_zep      | bl rx    _ |   2 |  8192 (  992) |  0x8099dc0 |  0x809bc34 
	  9 | RPL                  | bl rx    _ |   5 |  8192 ( 1080) |  0x809e7c0 |  0x80a0634 
	    | SUM                  |            |     | 90112 (12324)
> ifconfig
ifconfig
Iface  7   HWaddr: 3e:90:48:a8:59:e6 
           
           MTU:1500  HL:64  
           Source address length: 6
           Link type: wired
           inet6 addr: ff02::1/128  scope: local [multicast]
           inet6 addr: fe80::3c90:48ff:fea8:59e6/64  scope: local
           inet6 addr: ff02::1:ffa8:59e6/128  scope: local [multicast]
           inet6 addr: ff02::1a/128  scope: local [multicast]
           
           Statistics for Layer 2
            RX packets 0  bytes 0
            TX packets 1 (Multicast: 1)  bytes 64
            TX succeeded 1 errors 0
           Statistics for IPv6
            RX packets 0  bytes 0
            TX packets 1 (Multicast: 1)  bytes 50
            TX succeeded 1 errors 0

```

It would be great to have this for the release!

miri64 commented:
Have you cleaned between builds?

PeterKietzmann commented:
Sorry, my bad. I didn't appy your patch but just added ` USEMODULE += socket_zep` to the Makefile.dep and forget to take out `netdev2_tap`. However, it seems there are some parameters needed to run a networking application with `sock_zep`, right? Could you provide a hint about how to run it? Would make testing easier for others.

miri64 commented:
> However, it seems there are some parameters needed to run a networking application with sock_zep, right? Could you provide a hint about how to run it? Would make testing easier for others.

I was hoping the online help (-h) would be enough to give some hints, but `-zla` sets the local IP address for the socket, `-zlp` the local port, `-zra` the remote address and `-zrp` the remote address. For testing I recommend `-zla :: -zlp 17754 -zra ::1 -zrp 17754`. I'm not planning to document this officially yet, since I need to redo these parameters considering #6337 (which I hope will be merged this week) anyways.

PeterKietzmann commented:
Well, of course I have seen the online help, default values as well as the python run script. But I couldn't get it running out of the box. That's why I asked for a hint. To be more precise: `make term -zla :: -zlp 17754 -zra ::1 -zrp 17754` does not run the program on my side.

miri64 commented:
> Well, of course I have seen the online help, default values as well as the python run script. But I couldn't get it running out of the box. That's why I asked for a hint. To be more precise: make term -zla :: -zlp 17754 -zra ::1 -zrp 17754 does not run the program on my side.

Try

```bash
TERMFLAGS="-zla :: -zlp 17754 -zra ::1 -zrp 17754" make term
```

;-)

PeterKietzmann commented:
I succesfully sent UDP(/IPv6/6LoWPAN/802154/ZEP) packets between different native instances and sniffed with wireshark. IMO it would be great if usage would be "documented" somewhere. E.g. I personally wasn't aware of `TERMFLAGS`. ZEP parameters to communicate between instances and the wireshark interface to sniff could be written down. I can't provide a full review but hopefully this already helps.

miri64 commented:
[`TERMFLAGS` are already documented](http://doc.riot-os.org/getting-started.html#building-and-executing-an-example) (not very visible though, but that's the case for many other make macros either). I hope that with #6337 I'm able to make this whole thing a little bit easier to work with, so that's mainly why I held back on proper documentation for now.

miri64 commented:
Regarding sniffing: though it is a nice addition I don't see it as the main feature of this module. It also depends very much on what the ZEP parameters are (`lo` if destination is the loopback address `::1`, any other interface for any other destination address, can be bound to an interface by using the local address, etc.). As long as this module is in the experimental state it is documentation of this kind might only cause misunderstood usage of the module.

miri64 commented:
> I can't provide a full review but hopefully this already helps.

@OlegHahm @kaspar030 @LudwigKnuepfer anyone of you in for a review?

miri64 commented:
That was weird!??!

miri64 commented:
Rebased for now, but did not adapt new paremeter handling yet. Will adapt with less beer around my brain ;-)

miri64 commented:
Rebased again and adapted for both new option parsing and multi-interface support. Will provide test setup in a bit

miri64 commented:
Since we now support multiple interfaces on native, I adapted the border router accordingly (had no time to properly test yet, though, it runs and shows two interfaces :+1:). If you want to test with `gnrc_networking` you need to apply a patch again, but the `gnrc_border_router` example might not work with it (and we might need to adapt the default configuration for native in the future to avoid that):

```diff
diff --git a/boards/native/Makefile.dep b/boards/native/Makefile.dep
index f356492..30446af 100644
--- a/boards/native/Makefile.dep
+++ b/boards/native/Makefile.dep
@@ -1,5 +1,5 @@
 ifneq (,$(filter netdev_default gnrc_netdev_default,$(USEMODULE)))
-  USEMODULE += netdev2_tap
+  USEMODULE += socket_zep
 endif
 
 ifneq (,$(filter socket_zep,$(USEMODULE)))
diff --git a/boards/native/Makefile.include b/boards/native/Makefile.include
index a339978..7e33426 100644
--- a/boards/native/Makefile.include
+++ b/boards/native/Makefile.include
@@ -89,7 +89,7 @@ export LINKFLAGS += -ffunction-sections
 
 # set the tap interface for term/valgrind
 ifneq (,$(filter netdev_default gnrc_netdev_default,$(USEMODULE)))
-       export PORT ?= tap0
+       export PORT ?= -z [::1]:17754
 else
        export PORT =
 endif
```

If you want to use a different configuration for the ZEP socket, use 

```shell
PORT="-z [<local_addr>:<local_port>],<remote_addr>:<remote_port>" make term
```

When running the terminal.

miri64 commented:
Ping?

miri64 commented:
Rebased and adapted to current master (and squashed, because everything else would have been to complicated)

miri64 commented:
Fixed rebase error.

miri64 commented:
Fixed static errors

jia200x commented:
awesome! I'm on it ;)

miri64 commented:
Rebased to current master

miri64 commented:
Rebased to current master.

PeterKietzmann commented on cpu/native/include/socket_zep.h:
```diff
@@ -0,0 +1,83 @@
+extern "C" {
+#endif
+
+#define SOCKET_ZEP_FRAME_PAYLOAD_LEN    (116)   /**< maximum possible payload size */
```
Why?

miri64 replied:
I guess there is a reason, I just don't remember anymore :(

miri64 replied:
(but iirc this is arbitrary anyway)

PeterKietzmann replied:
But imo it should behave as an 802.15.4 radio device

PeterKietzmann commented on cpu/native/socket_zep/socket_zep.c:
```diff
@@ -0,0 +1,441 @@
+    .set = _set,
+};
+
+static size_t _zep_hdr_fill_v2_data(socket_zep_t *dev, zep_v2_data_hdr_t *hdr,
```
Why `v2`? Not just here

miri64 replied:
There are two versions of ZEP. This implements mostly v2 ;-).

PeterKietzmann commented on cpu/native/socket_zep/socket_zep.c:
```diff
@@ -0,0 +1,441 @@
+    if (netdev->event_callback) {
+        dev->last_event = NETDEV_EVENT_TX_STARTED;
+        netdev->event_callback(netdev, NETDEV_EVENT_ISR);
+        thread_yield();
```
Why do you need to `yield` here?

miri64 replied:
Don't know either... while testing this I realized this PR isn't working at the moment. Will look into it :(

PeterKietzmann replied:
Ok 'good' to hear. I didn't get it running either. Please ping me once fixed. Regarding `thread_yield()` I am wondering to which thread you want to yield...

PeterKietzmann commented on cpu/native/socket_zep/socket_zep.c:
```diff
@@ -0,0 +1,441 @@
+                DEBUG("socket_zep::recv: invalid ZEP header");
+                return -1;
+            }
+            switch (tmp->version) {
```
What's the benefit of the switch-case here?

miri64 replied:
guess https://github.com/RIOT-OS/RIOT/pull/6121#discussion_r119191851 invalidates this comment?

PeterKietzmann commented on sys/include/net/zep.h:
```diff
@@ -0,0 +1,103 @@
+} zep_hdr_t;
+
+/**
+ * @brief   ZEPv1 header definition
```
Now I understand that ZEPv1 vs ZEPv2 is not only a question of the number of iterations in RIOT. Now that v2 is implemented, shouldn't we remove definitions for v1 to avoid confusion?

miri64 replied:
I think it's more confusing to not provide that information.

miri64 replied:
(couldn't you have removed the related comments after you realized before posting ^^)

miri64 commented:
>  I still don't feel like being the only one looking at the changes in cpu/native.

There are no changes to cpu/native but in `startup.c` which is just straight forward parameter parsing. But if you do not feel comfortable, @LudwigKnuepfer may I get your blessing?

miri64 commented:
Just heard @cladmi knows some stuff about ZEP, so I assigned him to review.

miri64 commented:
Rebased to current master.

smlng commented:
murdock isn't all that happy, can you have look?

miri64 commented:
@smlng Murdock should now be happy.

smlng commented:
[Nope](https://ci.riot-os.org/RIOT-OS/RIOT/6121/ca1a23f25112ce2bf5bc165c6afa936066460a49/output/compile/examples/gnrc_border_router/native.txt)

miri64 commented:
That's a new one :D

miri64 commented:
Rebased (to include `gnrc_netif` to merge-base) and fixed.

miri64 commented:
@smlng now finally both Travis and Murdock are happy (except for squashing notice). Do you think we can include this into the release? Would be awesome for testing and debugging to have IEEE 802.15.4 for `native`.

BTW, @aabadie @kaspar030 @smlng Murdock doesn't seem to have `flake8` to be installed.

smlng commented:
> BTW, @aabadie @kaspar030 @smlng Murdock doesn't seem to have flake8 to be installed.

yep, workers need an update but not before release testing is done.

Otherwise, yes this should go into the release.

smlng commented on cpu/native/include/socket_zep_params.h:
```diff
@@ -0,0 +1,54 @@
+ * @{
+ *
+ * @file
+ * @brief       Configuration parameters for the @ref netdev_socket_zep driver
```
indention off? One tab too much?

smlng commented on cpu/native/include/socket_zep.h:
```diff
@@ -0,0 +1,83 @@
+/**
+ * @defgroup    netdev_socket_zep  ZEP-based IEEE 802.15.4 device for native
+ * @ingroup     netdev
+ * @brief
```
mhm, a bit sad this is empty 😢 Maybe put the (rather long) defgroup title here, and shorten the above?

smlng commented on sys/auto_init/netif/auto_init_socket_zep.c:
```diff
@@ -0,0 +1,61 @@
+#include "debug.h"
+
+/**
+ * @brief   Define stack parameters for the MAC layer thread
```
/brief/name/ for group

miri64 replied:
Isn't generated anyways and the other devices use the same notation (also the brace doesn't seem to be closed... I just remove the brace, because this just refers to the stack, not the prio.

smlng commented on sys/include/net/zep.h:
```diff
@@ -0,0 +1,103 @@
+/**
+ * @defgroup    net_zep  ZEP header definitions
+ * @ingroup     net
+ * @brief
```
group description empty again 😢

smlng commented:
while the error above was with `make term` the following comes from `make test`:

```
Socket ZEP device driver test
Initializing socket ZEP with (local: [::]:12345, remote: [::1]:17754)
(Hwaddrs: 3038, 005a45501c1e3038)
Send zero-length packet
Send 'Hello\0World\0'
Waiting for an incoming message (use `make test`)
Received invalid packet
Timeout in expect script at "child.expect(r"RSSI: \d+, LQI: \d+, Data:")" (tests/socket_zep/tests/01-run.py:50)
  File "/RIOT/dist/tools/testrunner/testrunner.py", line 55, in run
    testfunc(child)
  File "./tests/01-run.py", line 50, in testfunc
    child.expect(r"RSSI: \d+, LQI: \d+, Data:")
  File "/usr/local/lib/python3.6/site-packages/pexpect/spawnbase.py", line 321, in expect
    timeout, searchwindowsize, async)
  File "/usr/local/lib/python3.6/site-packages/pexpect/spawnbase.py", line 345, in expect_list
    return exp.expect_loop(timeout)
  File "/usr/local/lib/python3.6/site-packages/pexpect/expect.py", line 107, in expect_loop
    return self.timeout(e)
  File "/usr/local/lib/python3.6/site-packages/pexpect/expect.py", line 70, in timeout
    raise TIMEOUT(msg)

Run tests failed
make: *** [test] Error 1
```

miri64 commented:
Did you used the most current version? I literally just fixed that in the latest batch and at least on my machine it was working with that.

smlng commented:
> Did you used the most current version

okay, I'll quickly test again

smlng commented:
I'll verified and I'm on the latest commit, but still get the errors as reported above. Though, I've to admit I test on macOS ...

smlng commented:
retested on linux: works! So issues reported are likely macOS specific, still would be great to have it compatible in the future, if possible.

smlng commented:
please squash in a _timely_ manner, the feature freeze is approaching slowly 😉

smlng commented:
ACK and GO!

miri64 commented:
Awesome! :tada:
